internal/util/compare/equal.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package compare contains utils to compare objects.
+package compare
+
+import (
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+)
+
+// Diff uses cmp.Diff to compare x and y.
+// If cmp.Diff panics, Equal returns an error.
+func Diff(x any, y any) (equal bool, diff string, matchErr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			equal = false
+			if err, ok := r.(error); ok {
+				matchErr = err
+				return
+			}
+
+			if errMsg, ok := r.(string); ok {
+				matchErr = errors.New(errMsg)
+			}
```

Comment:
Should we add something like "unable to recover" in case of line 38 is `!ok`?


```suggestion
        return
			}
      
      matchErr = errors.New("unable to recover from cmp.Diff panic")
```

Reply:
Good catch. I wouldn't call this "not recovered", PTAL what I changed